### PR TITLE
Replace old packers

### PR DESCRIPTION
**Pull request by ceyonur**
## Why this should be merged

Initially we wrote our own Unpack/Pack methods and had a fixed length check for input bytes. In #487 we became aware that Solidity/JS libraries can add some extra padding bytes at the end of the input (input must still be multiple of 32). 

This PR replaces our legacy Packers with ones from ABI package to provide a better support. Also we have removed
the input len check with the DUpgrade to support Solidity padding. But we still need to keep this check before DUpgrade for backwards compatibility.

Closes: #510 #487 

Test branch #843 

## How this was tested

- [x] Unit tests
- [x] Local network testing

## How is this documented

This makes the code on-par with recent Solidity docs.

### CODE REVIEW

File: precompile/contracts/feemanager/contract.go [hunk 1]
```diff
@@ -239,6 +156,53 @@ func StoreFeeConfig(stateDB contract.StateDB, feeConfig commontype.FeeConfig, bl
...
+	// However solidity does not always pack the input to the correct length, and allows
+	// for extra padding bytes to be added to the end of the input. Therefore, we have removed
+	// this check with the DUpgrade. We still need to keep this check for backwards compatibility.
+	if !skipLenCheck && len(input) != feeConfigInputLen {
+		return commontype.FeeConfig{}, fmt.Errorf("%w: %d", ErrInvalidLen, len(input))
+	}
+	inputStruct := FeeConfigABIStruct{}
+	err := FeeManagerABI.UnpackInputIntoInterface(&inputStruct, "setFeeConfig", input)
+
+	if err != nil {
```
Comment (darioush) [inline precompile/contracts/feemanager/contract.go:188]:
```suggestion
	err := FeeManagerABI.UnpackInputIntoInterface(&inputStruct, "setFeeConfig", input)
	if err != nil {
```

File: precompile/contracts/nativeminter/module.go [hunk 2]
```diff
@@ -15,12 +15,13 @@ import (
 
 var _ contract.Configurator = &configurator{}
 
-// ConfigKey is the key used in json config files to specify this precompile config.
+// ConfigKey is the key used in json config files to specify this precompile precompileconfig.
```
Comment (darioush) [inline precompile/contracts/nativeminter/module.go:18]: seems this didn't need to be changed? could we instead update the template to either `precompile config` or
`precompileconfig` and also make all the existing precompiles uniform?

File: precompile/contracts/feemanager/config.go [hunk 3]
```diff
@@ -46,6 +46,8 @@ func NewDisableConfig(blockTimestamp *uint64) *Config {
 	}
 }
 
+// Key returns the key for the FeeManager precompileconfig.
+// This should be the same key as used in the precompile module.
```
Comment (darioush) [inline precompile/contracts/feemanager/config.go:50]: could we add these to deployerallowlist and txallowlist too?

**Review by darioush — APPROVED** [review]
just some comments about the comments

File: precompile/contracts/nativeminter/module.go (hunk 2 above)
Comment (ceyonur) [reply precompile/contracts/nativeminter/module.go]: fixed

File: precompile/contracts/feemanager/config.go (hunk 3 above)
Comment (ceyonur) [reply precompile/contracts/feemanager/config.go:50]: added

**Review by aaronbuchwald — CHANGES_REQUESTED** [review]
Could we add serialization test vectors to make sure that the precompile function arguments stays the same when they are packed and unpacked, so we can run the same test vector on the old code as well?

File: precompile/contracts/feemanager/unpack_pack_test.go [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+		// otherwise the value will be truncated when packed,
+		// and thus unpacked output will not be equal to the value
+		if bigIntVal.Cmp(abi.MaxUint256) > 0 {
+			doCheckOutputs = false
+		}
+		testOldPackGetFeeConfigOutputEqual(t, feeConfig, doCheckOutputs)
+	})
+}
+
+func TestPackUnpackGetFeeConfigOutputEdgeCases(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply precompile/contracts/feemanager/unpack_pack_test.go]: changed

File: precompile/contracts/feemanager/unpack_pack_test.go [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+	testOldPackGetFeeConfigOutputEqual(t, testFeeConfig, true)
+	// These should panic
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply precompile/contracts/feemanager/unpack_pack_test.go]: removed

File: precompile/contracts/nativeminter/unpack_pack_test.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+	f.Add(testAddrBytes, common.Big0.Bytes())
+	f.Add(testAddrBytes, common.Big1.Bytes())
+	f.Add(testAddrBytes, abi.MaxUint256.Bytes())
+	f.Add(testAddrBytes, new(big.Int).Sub(abi.MaxUint256, common.Big1).Bytes())
+	f.Add(testAddrBytes, new(big.Int).Add(abi.MaxUint256, common.Big1).Bytes())
+	f.Add(constants.BlackholeAddr.Bytes(), common.Big2.Bytes())
+	f.Fuzz(func(t *testing.T, b []byte, bigIntBytes []byte) {
+		bigIntVal := new(big.Int).SetBytes(bigIntBytes)
+		doCheckOutputs := true
+		// we can only check if outputs are correct if the value is less than MaxUint256
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply precompile/contracts/nativeminter/unpack_pack_test.go:37]: Currently the packer uses 256 bit two's complement number for bigints and then uses the first 32 bytes of the byte representation, which truncates if big int > maxuint256. I will keep investigating individual type packing in this issue https://github.com/ava-labs/subnet-evm/issues/769

File: precompile/contracts/nativeminter/contract_test.go [hunk 7]
```diff
@@ -9,156 +9,205 @@ import (
...
-				allowlist.TestEnabledAddr: math.NewHexOrDecimal256(2),
+				return input
 			},
+			SuppliedGas: MintGasCost,
+			ReadOnly:    false,
+			ExpectedErr: ErrCannotMint.Error(),
 		},
-		AfterHook: func(t testing.TB, state contract.StateDB) {
-			require.Equal(t, common.Big2, state.GetBalance(allowlist.TestEnabledAddr), "expected minted funds")
+		"calling mintNativeCoin from Enabled should succeed": {
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply precompile/contracts/nativeminter/contract_test.go:36]: I'd like to keep them specific as possible.

File: precompile/contracts/nativeminter/contract.go [hunk 8]
```diff
@@ -99,20 +107,27 @@ func mintNativeCoin(accessibleState contract.AccessibleState, caller common.Addr
...
-// is accessed controlled by an allow list at [precompileAddr].
+// createNativeMinterPrecompile returns a StatefulPrecompiledContract with getters and setters for the precompile.
+// Access to the getters/setters is controlled by an allow list for ContractAddress.
 func createNativeMinterPrecompile() contract.StatefulPrecompiledContract {
-	enabledFuncs := allowlist.CreateAllowListFunctions(ContractAddress)
+	var functions []*contract.StatefulPrecompileFunction
+	functions = append(functions, allowlist.CreateAllowListFunctions(ContractAddress)...)
 
-	mintFunc := contract.NewStatefulPrecompileFunction(mintSignature, mintNativeCoin)
+	abiFunctionMap := map[string]contract.RunStatefulPrecompileFunc{
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply precompile/contracts/nativeminter/contract.go:116]: My intention was to align all precompile implementations and precompile template.

File: precompile/contracts/feemanager/module.go [hunk 9]
```diff
@@ -20,6 +20,7 @@ const ConfigKey = "feeManagerConfig"
 
 var ContractAddress = common.HexToAddress("0x0200000000000000000000000000000000000003")
 
+// Module is the precompile module. It is used to register the precompile contract.
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply precompile/contracts/feemanager/module.go:23]: Precompile packages are designed to be standalone as possible without requiring any interference from outer packages or modification to other packages. This requires some redundancy in precompile packages but with this way precompiles can self-register themselves to the module registry. 

That being said, there is always room for more refactoring if you think we can reduce the boilerplate code.

File: precompile/contracts/feemanager/contract_test.go [hunk 10]
```diff
@@ -248,6 +271,58 @@ var (
 			ReadOnly:    false,
 			ExpectedErr: vmerrs.ErrOutOfGas.Error(),
 		},
+		"set config with extra padded bytes should fail before DUpgrade": {
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply precompile/contracts/feemanager/contract_test.go:274]: I think it's probably leave it as it is. It's a bit duplicated code but I prefer to use the same test framework without special casing.

File: precompile/contracts/nativeminter/module.go [hunk 11]
```diff
@@ -36,11 +37,14 @@ func init() {
 	}
 }
 
+// MakeConfig returns a new precompile config instance.
+// This is required for Marshal/Unmarshal the precompile config.
```
[the remark this answers is not in the document]
Comment (ceyonur) [reply precompile/contracts/nativeminter/module.go]: done

**Review by aaronbuchwald — APPROVED** [review]

File: precompile/contracts/nativeminter/unpack_pack_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ceyonur) [reply precompile/contracts/nativeminter/unpack_pack_test.go:37]: That's a question for #769. We currently use ABI out of package (with a few additions).

**Review by darioush — APPROVED** [review]

**Pull request merged**